R/download.R

```diff
@@ -4,6 +4,104 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
+#' Download Bugzilla issues using the Bugzilla REST API
+#'
+#' Downloads bugzilla issues into a folder, where each file is a json containing a page of issues.
+#' Returns a list of bugzilla bug ids for use with \code{\link{download_bugzilla_comments_from_rest_api}}.
+#'
+#' @param bugzilla_site URL to specific bugzilla site
+#' @param start_timestamp when to start bug retrieval (ex. 2023-01-01T00:14:57Z)
+#' @param save_folder_path the full *folder* path where the bugzilla issues will be stored
+#' @param limit the number of issues saved in each page file. Some bugzilla sites have limits set on how many bugs
+#' can be retrieved in one GET request, in which case, the limit should be set accordingly.
+#' @seealso \code{\link{download_bugzilla_comments_from_rest_api}} and \code{\link{parse_bugzilla_crawler_issue}}
+#' @export
+download_bugzilla_issues_from_rest_api <- function(bugzilla_site, start_timestamp, save_folder_path, limit=500){
+  # Format link to retrieve data using Bugzilla REST API
+  bugzilla_site <- paste(bugzilla_site, "/rest", sep="")
+
+  # Make sure folder path is correctly formatted
+  if (stringi::stri_sub(save_folder_path,-1) != "/"){
+    save_folder_path <-paste0(save_folder_path, "/")
+  }
+
+  # Holds the bug ids
+  bug_ids <- c()
+  # Defines what bug to start from in bugs retrieved.
+  offset <- 0
+  # Defines name of the file. Each page contains 500 bugs.
+  page <- 0
+
+  # Get the issues starting from the specified date
+  issues <- httr::GET(paste0(bugzilla_site, "/bug", "?creation_time=", start_timestamp, "&limit=", limit, "&offset=", offset))
+
+  # Save the issues and the bug ids if there are issues.
+  if (length(httr::content(issues)$bugs) > 0){
+    jsonlite::write_json(jsonlite::fromJSON(httr::content(issues, "text")),
+                         paste0(save_folder_path, page, ".json"), auto_unbox=TRUE)
+    json_issues <- rawToChar(httr::content(issues, as = "raw"))
+
+    # Get the ids of the bugs to download comments associated with these bugs in
+    # download_bugzilla_comments_from_rest_api
+    current_bug_ids <- jsonlite::fromJSON(json_issues)[["bugs"]][["id"]]
+    bug_ids <- c(bug_ids, current_bug_ids)
+    offset <- offset + limit
```

Comment:
My second question goes here.

> If a user was to specify a limit of 500, the first GET request would retrieve bugs 0-20, and the second would retrieve bugs 500-520, and so on, since I am using the limit to adjust the "offset" parameter, which determines which bugs to retrieve for each GET request.

Would it be correct to say that if you parsed only your first saved file to know how many issues were _actually_ retrieved, you would then be able to correct any user assumption on the limit, and safely execute the function?

I guess I am also wondering why you must rely on the "limit" and "offset" as a parameter.  With GitHub API, you are mainly iterating through pages, and there is an endpoint you can assume is equivalent to click the "Next" page to show. Is there no equivalent here?

Reply:
>How did you find this page? Via google? Or did their Bugzilla tracker had at the bottom of the page some link that led you there? If Bugzilla has a standard way to find these pages, it would be very helpful to who is using this function since you already went through this.

Hi Carlos, I found it on Google. I don't think there's a standard.

>Would it be correct to say that if you parsed only your first saved file to know how many issues were actually retrieved, you would then be able to correct any user assumption on the limit, and safely execute the function?

Yes, that's right. A user could wait until at least two files were created for the bugs they downloaded and check if the first json file has a limit that is under what the user specified. Then they could use the limit they see as the limit parameter. Or they could check the parsed table.

>I guess I am also wondering why you must rely on the "limit" and "offset" as a parameter. With GitHub API, you are mainly iterating through pages, and there is an endpoint you can assume is equivalent to click the "Next" page to show. Is there no equivalent here?

I have not found an equivalent. The offset & limit seems like the main means of pagination.

Reply:
> Yes, that's right. A user could wait until at least two files were created for the bugs they downloaded and check if the first json file has a limit that is under what the user specified. Then they could use the limit they see as the limit parameter. Or they could check the parsed table.

Actually, what I am asking is if your function can do that for the user instead. After the first page is done, could Malia parser be used to read one page of issues, assess it's length, and then use that as the `offset` thereafter? If so, the user `limit` parameter would be a `limit_upperbound`, and your function would always work.

Reply:
Got it, I will try that out.

Reply:
Yes, this works for me. Now that my function depends on Malia's, do you want me to add her parser to my pull request? Or should I wait for her parser pull request to be merged with main.

Reply:
Hmm, I'd say let's wait so her code goes through her and in case there are more revisions from her side! 

At this point, I'd suggest you git squash since there is some conflicts that need to be resolved before I can accept your PR. 

Also, please don't forget to expand your function body for the downloader to include the details of the limit parameter discussed here. They are very important, and I feel I will be asked this or ask it myself within a few months for forgetting! 

Also, please see the e-mail I just sent you :)